packages/babel-types/scripts/generators/typescript.js

```diff
@@ -109,25 +116,113 @@ for (const type in t.NODE_FIELDS) {
   }
 }
 
-for (let i = 0; i < t.TYPES.length; i++) {
-  let decl = `export function is${t.TYPES[i]}(node: object | null | undefined, opts?: object | null): `;
-
-  if (t.NODE_FIELDS[t.TYPES[i]]) {
-    decl += `node is ${t.TYPES[i]};`;
-  } else if (t.FLIPPED_ALIAS_KEYS[t.TYPES[i]]) {
-    decl += `node is ${t.TYPES[i]};`;
-  } else {
-    decl += `boolean;`;
-  }
-
-  lines.push(decl);
+for (const typeName of t.TYPES) {
+  const result =
+    t.NODE_FIELDS[typeName] || t.FLIPPED_ALIAS_KEYS[typeName]
+      ? `node is ${typeName}`
+      : "boolean";
+
+  lines.push(
+    `export function is${typeName}(node: object | null | undefined, opts?: object | null): ${result};`,
+    // TypeScript 3.7: https://github.com/microsoft/TypeScript/pull/32695 will allow assert declarations
+    // eslint-disable-next-line max-len
+    `// export function assert${typeName}(node: object | null | undefined, opts?: object | null): asserts ${
+      result === "boolean" ? "node" : result
+    };`
+  );
 }
 
 lines.push(
```

Comment:
I think this whole thing would be much nicer to modify if there was, say, a `typescript.template.d.ts` file that is complete enough to be typechecked and with placeholder comments / types for where the generated types should get injected; the generator script then just needs to slurp the template and edit it.